### PR TITLE
Removing Id Usages From CLI and CLI Tests packages

### PR DESCRIPTION
**Pull request by knazir**
Phasing out deprecated Id usages in CLI and CLI Tests packages.

### CODE REVIEW

**Comment by knazir** [comment]
Build running: http://builds.cask.co/browse/CDAP-DUT4796-2

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java [hunk 1]
```diff
@@ -101,15 +105,15 @@
...
-  private final Id.Workflow fakeWorkflowId = Id.Workflow.from(fakeAppId, FakeWorkflow.NAME);
-  private final Id.Flow fakeFlowId = Id.Flow.from(fakeAppId, FakeFlow.NAME);
-  private final Id.Program fakeSparkId = Id.Program.from(fakeAppId, ProgramType.SPARK, FakeSpark.NAME);
-  private final Id.Service pingServiceId = Id.Service.from(fakeAppId, PingService.NAME);
-  private final Id.Service prefixedEchoHandlerId = Id.Service.from(fakeAppId, PrefixedEchoHandler.NAME);
-  private final Id.DatasetInstance fakeDsId = Id.DatasetInstance.from(Id.Namespace.DEFAULT, FakeApp.DS_NAME);
-  private final Id.Stream fakeStreamId = Id.Stream.from(Id.Namespace.DEFAULT, FakeApp.STREAM_NAME);
+  private final ArtifactId fakeArtifactId = NamespaceId.DEFAULT.artifact(FakeApp.NAME, "1.0");
+  private final ApplicationId fakeAppId = NamespaceId.DEFAULT.app(FakeApp.NAME);
+  private final ProgramId fakeWorkflowId = fakeAppId.workflow(FakeWorkflow.NAME);
```
Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java:111]: this will need to be rebased with #6682

File: cdap-cli/pom.xml [hunk 2]
```diff
@@ -72,6 +72,10 @@
       <artifactId>concurrent-trees</artifactId>
     </dependency>
     <dependency>
+      <groupId>org.apache.twill</groupId>
```
Comment (bdmogal) [inline cdap-cli/pom.xml:76]: why is this needed?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/CallServiceCommand.java [hunk 3]
```diff
@@ -96,7 +96,8 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
     }
 
     Map<String, String> headerMap = GSON.fromJson(headers, new TypeToken<Map<String, String>>() { }.getType());
-    URL url = new URL(serviceClient.getServiceURL(service), path);
+    URL url = new URL(serviceClient.getServiceURL(Id.Service.from(service.getParent().toId(), service.getEntityName())),
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/CallServiceCommand.java]: cant do `(Id.Service) service.toId`?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/CheckServiceAvailabilityCommand.java [hunk 4]
```diff
@@ -54,8 +55,9 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
 
     String appId = appAndServiceId[0];
     String serviceName = appAndServiceId[1];
-    Id.Service serviceId = Id.Service.from(cliConfig.getCurrentNamespace(), appId, serviceName);
-    output.println(serviceClient.getAvailability(serviceId));
+    ProgramId serviceId = cliConfig.getCurrentNamespace().app(appId).service(serviceName);
+    output.println(serviceClient.getAvailability(Id.Service.from(serviceId.getParent().toId(),
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/CheckServiceAvailabilityCommand.java]: ditto

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetServiceEndpointsCommand.java [hunk 5]
```diff
@@ -60,8 +61,9 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
 
     String appId = appAndServiceId[0];
     String serviceName = appAndServiceId[1];
-    Id.Service serviceId = Id.Service.from(cliConfig.getCurrentNamespace(), appId, serviceName);
-    List<ServiceHttpEndpoint> endpoints = serviceClient.getEndpoints(serviceId);
+    ProgramId serviceId = cliConfig.getCurrentNamespace().app(appId).service(serviceName);
+    List<ServiceHttpEndpoint> endpoints = serviceClient.getEndpoints(Id.Service.from(serviceId.getParent().toId(),
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/GetServiceEndpointsCommand.java]: ditto

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetWorkflowLocalDatasetsCommand.java [hunk 6]
```diff
@@ -59,7 +59,7 @@ public void perform(Arguments arguments, PrintStream printStream) throws Excepti
       throw new CommandInputError(this);
     }
 
-    ProgramRunId programRunId = new ProgramRunId(cliConfig.getCurrentNamespace().getId(), programIdParts[0],
+    ProgramRunId programRunId = new ProgramRunId(cliConfig.getCurrentNamespace().getNamespace(), programIdParts[0],
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/GetWorkflowLocalDatasetsCommand.java]: will `cliConfig.getCurrentNamespace().app(programIdParts[0]).workflow(programIdParts[1]).run(arguments.get(ArgumentName.RUN_ID.toString())` be better?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetWorkflowStateCommand.java [hunk 7]
```diff
@@ -61,7 +61,7 @@ public void perform(Arguments arguments, PrintStream printStream) throws Excepti
       throw new CommandInputError(this);
     }
 
-    ProgramRunId programRunId = new ProgramRunId(cliConfig.getCurrentNamespace().getId(), programIdParts[0],
+    ProgramRunId programRunId = new ProgramRunId(cliConfig.getCurrentNamespace().getNamespace(), programIdParts[0],
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/GetWorkflowStateCommand.java]: ditto

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/app/RestartProgramsCommand.java [hunk 8]
```diff
@@ -49,14 +49,14 @@ protected BatchProgram createProgram(ProgramRecord programRecord) {
   @Override
   protected void runBatchCommand(PrintStream printStream, Args<BatchProgram> args) throws Exception {
     printStream.print("Stopping programs...\n");
-    programClient.stop(args.appId.getNamespace(), args.programs);
+    programClient.stop(args.appId.getParent().toId(), args.programs);
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/app/RestartProgramsCommand.java]: nit: store `args.appId.getParent().toId()` in a variable so you don't need to create extra objects on line 59

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/artifact/LoadArtifactCommand.java [hunk 9]
```diff
@@ -55,34 +55,35 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
...
     } else {
-      artifactId = Id.Artifact.from(cliConfig.getCurrentNamespace(), name, version);
+      artifactId = cliConfig.getCurrentNamespace().artifact(name, version);
     }
 
     String configPath = arguments.getOptional(ArgumentName.ARTIFACT_CONFIG_FILE.toString());
     if (configPath == null) {
-      artifactClient.add(artifactId.getNamespace(), artifactId.getName(), Files.newInputStreamSupplier(artifactFile),
-                         artifactId.getVersion().getVersion());
+      artifactClient.add(artifactId.getParent().toId(), artifactId.getEntityName(),
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/artifact/LoadArtifactCommand.java]: better to extract `artifactId.getParent().toId()` to a variable.

File: cdap-cli/src/main/java/co/cask/cdap/cli/completer/element/HttpEndpointPrefixCompleter.java [hunk 10]
```diff
@@ -60,19 +61,19 @@ public int complete(String buffer, int cursor, List<CharSequence> candidates) {
...
+        programIdArgument.getProgramId());
       completer.setEndpoints(getEndpoints(service, arguments.get(METHOD)));
     } else {
       completer.setEndpoints(Collections.<String>emptyList());
     }
     return super.complete(buffer, cursor, candidates);
   }
 
-  public Collection<String> getEndpoints(Id.Service serviceId, String method) {
+  public Collection<String> getEndpoints(ProgramId serviceId, String method) {
```
Comment (bdmogal) [inline cdap-cli/src/main/java/co/cask/cdap/cli/completer/element/HttpEndpointPrefixCompleter.java]: hmmm perhaps we do need a `ServiceId`. @anwar6953 ?

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java [hunk 11]
```diff
@@ -47,6 +48,28 @@ public ArtifactId artifact(String artifact, String version) {
     return new ArtifactId(namespace, artifact, version);
   }
 
+  public ArtifactId parseArtifact(String fileName) {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java]: is this copied from somewhere?

**Review by bdmogal — COMMENTED** [review]
@knazir mostly minor comments.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/CallServiceCommand.java (hunk 3 above)
Comment (knazir) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/CallServiceCommand.java]: Tried that but can't cast a `ProgramId` to `Id.Service`.

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/CheckServiceAvailabilityCommand.java (hunk 4 above)
Comment (knazir) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/CheckServiceAvailabilityCommand.java]: ditto

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetServiceEndpointsCommand.java (hunk 5 above)
Comment (knazir) [reply cdap-cli/src/main/java/co/cask/cdap/cli/command/GetServiceEndpointsCommand.java]: ditto

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/GetProgramInstancesCommand.java [hunk 12]
```diff
@@ -56,24 +59,24 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
...
         break;
       case SERVICE:
         if (programIdParts.length < 2) {
           throw new CommandInputError(this);
         }
         String serviceName = programIdParts[1];
-        Id.Service service = Id.Service.from(appId, serviceName);
-        instances = programClient.getServiceInstances(service);
+        ProgramId service = appId.service(serviceName);
+        instances = programClient.getServiceInstances((Id.Service) service.toId());
```
Comment (knazir) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/GetProgramInstancesCommand.java]: fix this

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java (hunk 11 above)
Comment (knazir) [reply cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java]: Yep, from `Id.Artifact`

File: cdap-cli/pom.xml (hunk 2 above)
Comment (anwar6953) [reply cdap-cli/pom.xml:76]: why is this needed

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java (hunk 11 above)
Comment (anwar6953) [reply cdap-proto/src/main/java/co/cask/cdap/proto/id/NamespaceId.java]: Why is it in `NamespaceId`, then? Why not in `ArtifactId`?

**Comment by anwar6953** [comment]
Just two comments:
https://github.com/caskdata/cdap/pull/6710#r79465189
https://github.com/caskdata/cdap/pull/6710#r79467023

File: cdap-cli/pom.xml (hunk 2 above)
Comment (bdmogal) [reply cdap-cli/pom.xml:76]: This is needed because `ProgramId` has a dependency on Twill's `RunId` due to the `run(RunId runId)` method.

Comment (bdmogal) [reply cdap-cli/pom.xml:76]: Added a comment explaining this.

**Comment by bdmogal** [comment]
@anwar6953 addressed comments in the last few commits. please review again.

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/ServiceId.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ * the License.
+ */
+
+package co.cask.cdap.proto.id;
+
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+
+/**
+ * Uniquely identifies a workflow.
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/ServiceId.java]: update javadoc to be accurate

**Comment by anwar6953** [comment]
lgtm besides the last comment about inaccurate javadoc

**Comment by bdmogal** [comment]
Addressed comments. Squashed commits. Waiting for http://builds.cask.co/browse/CDAP-DUT4796-8

**Comment by bdmogal** [comment]
Build succeeded. Merging.
